src/Functions/FunctionMapMapped.h

```diff
@@ -0,0 +1,198 @@
+                            + toString(arguments.size()) + ".",
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        if (arguments.size() == 1)
+            throw Exception("Function " + getName() + " needs at least one map argument.",
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        DataTypes nested_types((arguments.size() - 1) * 2);
+        for (size_t i = 0; i < arguments.size() - 1; ++i)
+        {
+            const DataTypeMap * map_type = checkAndGetDataType<DataTypeMap>(&*arguments[i + 1]);
+            if (!map_type)
+                throw Exception("Argument " + toString(i + 2) + " of function " + getName() + " must be map. Found "
+                                + arguments[i + 1]->getName() + " instead.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            nested_types[i * 2] = recursiveRemoveLowCardinality(map_type->getKeyType());
+            nested_types[i * 2 + 1] = recursiveRemoveLowCardinality(map_type->getValueType());
+        }
+
+        const DataTypeFunction * function_type = checkAndGetDataType<DataTypeFunction>(arguments[0].get());
+        if (!function_type || function_type->getArgumentTypes().size() != nested_types.size())
+            throw Exception("First argument for this overload of " + getName() + " must be a function with "
+                            + toString(nested_types.size()) + " arguments. Found "
+                            + arguments[0]->getName() + " instead.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        arguments[0] = std::make_shared<DataTypeFunction>(nested_types);
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        size_t min_args = Impl::needExpression() ? 2 : 1;
+        if (arguments.size() < min_args)
+            throw Exception("Function " + getName() + " needs at least "
+                            + toString(min_args) + " argument; passed "
+                            + toString(arguments.size()) + ".",
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        if (arguments.size() > 2 && Impl::needOneMap())
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function '{}' needs one map argument", getName());
+
+        const auto * data_type_function = checkAndGetDataType<DataTypeFunction>(arguments[0].type.get());
+
+        if (!data_type_function)
+            throw Exception("First argument for function " + getName() + " must be a function.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        /// The types of the remaining arguments are already checked in getLambdaArgumentTypes.
+
+        DataTypePtr return_type = removeLowCardinality(data_type_function->getReturnType());
+        if (Impl::needBoolean() && !WhichDataType(return_type).isUInt8())
+            throw Exception(
+                "Expression for function " + getName() + " must return UInt8, found " + return_type->getName(),
```

Comment:
We have `Bool` type now (but actually it's just an alias for UInt8) 
https://github.com/ClickHouse/ClickHouse/blob/master/src/DataTypes/DataTypeDomainBool.cpp